halo2-base/src/safe_types/mod.rs

```diff
@@ -84,6 +84,55 @@ pub type SafeAddress<F> = SafeType<F, 1, 160>;
 /// SafeType for bytes32.
 pub type SafeBytes32<F> = SafeType<F, 1, 256>;
 
+/// Represents a variable length byte array as a vector of AssignedValue<F>'s.
+/// 
+/// The number of elements within the array may vary in circuit from `0..max_var_len`, with the assigned witness `var_len` 
+/// representing in circuit the number of significant elements present within the byte array from `0..=var_len`.
+/// 
+/// * Range checks that each AssignedValue<F> of the vector is within byte range 0~255.
+/// * Asserts that bytes.len() == max_var_len and var_len < max_var_len.
+/// 
+/// Note: The minimum value of max_var_len (and length of the byte array) is 1 i.e. for var_len = 32, max_var_len must be 33.
+#[derive(Debug, Clone)]
+pub struct VariableByteArray<F: ScalarField> {
```

Comment:
Can max_var_len be passed as a const generic parameter? The idea is that the max size should be determined at compile time. If it's determined at run time, it's possible that its values are different between keygen and proof generation.

Reply:
Good idea!

Reply:
@nyunyunyunyu I don't think this will work from a dev ex perspective: the common use case is we want to store a `Vec` of `VarLenBytes`. If they all have different const generics this will be very hard to deal with. We can make the constructor `raw_bytes_to...` have a const generic in there to safeguard - but I need to check if this actually matches our usage.

Reply:
@jonathanpwang Read through the codebase and I think I might understand your concern. Correct me if I'm wrong:
In `axiom-eth`, we use `Vec<Vec<AssignedValue<F>>>` in many places. There are actually 2 types:

1. [MPTFixedKeyProof](https://github.com/axiom-crypto/axiom-eth/blob/e032a2447be0cc471af0d290dda2843657a00f8b/axiom-eth/src/mpt/mod.rs#L154)/[MPTFixedKeyProofWitness](https://github.com/axiom-crypto/axiom-eth/blob/e032a2447be0cc471af0d290dda2843657a00f8b/axiom-eth/src/mpt/mod.rs#L173)
- In these places, as mentioned in comments, we actually want to enforce the max length constrain becuase these structs **owns these fields** which are assigned by this crate. 
- With the const generic parameters, we can easily know the shape of each field. Codes will be more readable.
- In a `Vec`, all elements should have same shape. If we really need elements with a different shape, put them into another field.
- Even if there is really a scenario(I cannot come up with one now) that we need tens of elements in differnt shapes, marcos can help us to acheive it.

2. [KeccakChip](https://github.com/axiom-crypto/axiom-eth/blob/e032a2447be0cc471af0d290dda2843657a00f8b/axiom-eth/src/keccak/mod.rs#L99) (I guess it's one of the places to "store a `Vec` of `VarLenBytes`")
- `KeccakChip` doesn't own or assign these cells. Everything in `var_len_queries`/`fixed_len_queries` is more like pointers. `KeccakChip` simply copies these cells to its own columns then produces outputs.
- If we just want to read witnesses of `SafeType`, I'm comfortable with opening `SafeType`. It's totally valid for a type to take assumption of `SafeType` and wrap its witnesses as a new semantic type. We can call `bytes(&self)` in `VariableByteArray` and wrap the return value as a `KeccakFixedLenQuery` or `KeccakVarLenQuery`.